setup_build.py

```diff
@@ -117,22 +117,27 @@ def _make_extensions(cls, config):
         if os.name != 'nt':
             settings['runtime_library_dirs'] = settings['library_dirs']
 
-        return [cls._make_extension(m, settings) for m in MODULES]
+        # 'NPY_1_21_API_VERSION' == 0x0000000e
+        settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_1_21_API_VERSION'))
+        extensions = [cls._make_extension(m, settings) for m in MODULES]
+
+        if int(numpy.__version__.split('.')[0]) >= 2:
+            # Enable NumPy 2.0 C API for modules that require it.
+            # NUMPY2_MODULES will not be importable when NumPy 1.x is installed.
+            # 'NPY_2_0_API_VERSION' == 0x00000012
+            settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_2_0_API_VERSION'))
+            extensions.extend(cls._make_extension(m, settings) for m in MODULES_NUMPY2_ONLY)
```

Comment:
```suggestion
        extensions = [cls._make_extension(m, settings) for m in MODULES]

        if int(numpy.__version__.split('.')[0]) >= 2:
            # Enable NumPy 2.0 C API for modules that require it.
            # MODULES_NUMPY2_ONLY will not be importable when NumPy 1.x is installed.
            # 'NPY_2_0_API_VERSION' == 0x00000012
            settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_2_0_API_VERSION'))
            extensions.extend(cls._make_extension(m, settings) for m in MODULES_NUMPY2_ONLY)
        else:
                # 'NPY_1_21_API_VERSION' == 0x0000000e
                settings['define_macros'].append(('NPY_TARGET_VERSION', 'NPY_1_21_API_VERSION'))

```

Reply:
Good point, better build all extensions with a consistent `NPY_TARGET_VERSION`. Not sure your suggestion applies as is though so I'll write my own version of it. Thanks !

Reply:
Actually this doesn't seem to work; the resulting wheels are not compatible with numpy 1.21, so I reverted this change.

https://github.com/h5py/h5py/actions/runs/16965494246/job/48088254143?pr=2646